modules/core/src/main/java/org/apache/synapse/config/xml/ForEachMediatorFactory.java

```diff
@@ -0,0 +1,130 @@
+
+    private static final QName FOREACH_Q =
+            new QName(SynapseConstants.SYNAPSE_NAMESPACE, "foreach");
+
+    private static final QName ID_Q
+            = new QName(XMLConfigConstants.NULL_NAMESPACE, "id");
+
+    public QName getTagQName() {
+        return FOREACH_Q;
+    }
+
+    @Override
+    protected Mediator createSpecificMediator(OMElement elem,
+                                              Properties properties) {
+        ForEachMediator mediator = new ForEachMediator();
+        processAuditStatus(mediator, elem);
+        OMAttribute expression = elem.getAttribute(ATT_EXPRN);
+
+        OMAttribute id = elem.getAttribute(ID_Q);
+        if (id != null) {
+            mediator.setId(id.getAttributeValue());
+        }
+
+        if (expression != null) {
+            try {
+                mediator.setExpression(SynapseXPathFactory.getSynapseXPath(elem,
+                        ATT_EXPRN));
+            } catch (JaxenException e) {
+                handleException("Unable to build the ForEach Mediator. " +
+                        "Invalid XPath " +
+                        expression.getAttributeValue(), e);
+            }
+        } else {
+            handleException("XPath expression is required "
+                    + "for an ForEach Mediator under the \"expression\" attribute");
+        }
+
+        OMAttribute sequenceAttr = elem.getAttribute(
+                new QName(XMLConfigConstants.NULL_NAMESPACE, "sequence"));
+        if (sequenceAttr != null && sequenceAttr.getAttributeValue() != null) {
+            mediator.setSequenceRef(sequenceAttr.getAttributeValue());
+        }
+
+        OMElement sequence = elem.getFirstChildWithName(
+                new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "sequence"));
+        if (sequence != null) {
+            SequenceMediatorFactory fac = new SequenceMediatorFactory();
+            SequenceMediator sequenceMediator = fac.createAnonymousSequence(sequence, properties);
+            if (validateSequence(sequenceMediator)) {
+                mediator.setSequence(sequenceMediator);
+            } else {
+                handleException("Sequence cannot contain Call, Send or CallOut mediators");
+            }
+        }
+        return mediator;
+    }
+
+    private boolean validateSequence(SequenceMediator sequence) {
+        boolean valid = true;
+        if (sequence != null) {
+            List<Mediator> mediators = sequence.getList();
+            for (Mediator m : mediators) {
+                if (m instanceof CallMediator) {
+                    valid = false;
+                    break;
+                } else if (m instanceof CalloutMediator) {
+                    valid = false;
+                    break;
+                } else if (m instanceof SendMediator) {
+                    valid = false;
+                    break;
+                }
+            }
+        }
+        return valid;
+    }
+
```

Comment:
Can be simplify like follows.
private boolean validateSequence(SequenceMediator sequence) {
        if (sequence != null) {
            List<Mediator> mediators = sequence.getList();
            for (Mediator m : mediators) {
                if (m instanceof CallMediator || m instanceof CalloutMediator ||
                    m instanceof SendMediator) {
                    return false;
                }
            }
        }
        return true;
    }